File: dox_development/dox_development/report/supplier_statement/supplier_statement.js
// Copyright (c) 2023, Nesscale Solutions Private Limited and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Supplier Statement"] = {
	"filters": [
		{
			"fieldname": "party",
			"label": __("Party"),
			"fieldtype": "Link",
			"options": "Supplier",
			"reqd": 1,
			get_data: function (txt) {
				if (!frappe.query_report.filters) return;

				let party_type = frappe.query_report.get_filter_value('party_type');
				if (!party_type) return;

				return frappe.db.get_link_options(party_type, txt);
			},
			on_change: function () {
				var party_type = "Supplier";
				var parties = frappe.query_report.get_filter_value('party');
				var party = parties;
				var fieldname = erpnext.utils.get_party_name(party_type) || "name";
				frappe.db.get_value(party_type, party, fieldname, function (value) {
					frappe.query_report.set_filter_value('party_name', value[fieldname]);
				});
				// if (!party_type || parties.length === 0 || parties.length > 1) {
				// 	frappe.query_report.set_filter_value('party_name', "");
				// 	return;
				// } else {

				// }
			}
		},
		{
			"fieldname": "party_name",
			"label": __("Party Name"),
			"fieldtype": "Read Only"
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 1,
			"width": "60px"
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1,
			"width": "60px"
		},
		{
			"fieldname": "presentation_currency",
			"label": __("Currency"),
			"fieldtype": "Link",
			"reqd": 1,
			"width": "60px",
			"options": "Currency",
		},
	]
};

File: dox_development/dox_development/report/supplier_statement/supplier_statement.py
# Copyright (c) 2023, Nesscale Solutions Private Limited and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from erpnext import get_default_company


def execute(filters=None):
    company = get_default_company()
    columns, data = [], []
    if not filters.get("from_date") or not filters.get("to_date"):
        frappe.throw(_("Select First from date and to date"))
        return columns, data
    if not filters.get("party"):
        frappe.throw(_("Select Supplier to view statement"))
        return columns, data
    filters_report = {
        "company": company,
        "from_date": filters.get("from_date"),
        "to_date": filters.get("to_date"),
        "party_type": "Supplier",
        "party": [filters.get("party")],
        "party_name": filters.get("party_name"),
        "group_by": "Group by Voucher (Consolidated)",
        "cost_center": [],
        "project": [],
        "include_dimensions": 1,
        "presentation_currency": filters.get("presentation_currency")
    }
    from frappe.desk.query_report import run

    res = run("General Ledger", filters=filters_report)
    data = res.get("result")
    columns = [
        {
            "label": "Posting Date",
            "fieldname": "posting_date",
            "fieldtype": "Date",
            "width": 120,
        },
        {"label": "Voucher Type", "fieldname": "voucher_type", "width": 120},
        {
            "label": "Voucher No",
            "fieldname": "voucher_no",
            "fieldtype": "Dynamic Link",
            "options": "voucher_type",
            "width": 180,
        },
        {
            "label": "Debit",
            "fieldname": "debit",
            "fieldtype": "Float",
            "width": 120,
        },
        {
            "label": "Credit",
            "fieldname": "credit",
            "fieldtype": "Float",
            "width": 120,
        },
        {
            "label": "Balance",
            "fieldname": "balance",
            "fieldtype": "Float",
            "width": 130,
        },
    ]
    return columns, data
